src/charm.py

```diff
@@ -227,13 +224,9 @@ def _on_collect_unit_status(self, event: ops.CollectStatusEvent) -> None:
         snap_service = get_installed_snap_service(SNAP_NAME)
 
         if not snap_service.present:
-            event.add_status(
-                BlockedStatus("snap service is not installed, please check snap service")
-            )
+            raise RuntimeError(f"{SNAP_NAME} snap is not installed")
         elif not snap_service.is_active():
-            event.add_status(
-                BlockedStatus("snap service is not running, please check snap service")
-            )
+            raise RuntimeError(f"{SNAP_NAME} snap service is not active")
```

Comment:
I am concerned this will not work well with `_on_collect_unit_status`

Reply:
I think `event.add_status` can work with ErrorStatus.

See: https://github.com/canonical/operator/blob/98a4e4f44d37d3361a253bc1eff49eee2ca6d140/ops/model.py#L1960-L1975

Reply:
(not verified) Still, it should not work because `ErrorStatus` has the highest priority. It means that install event -> collect status -> raise Error status. This will happen whether or not you remove the cos-agent relation check

Reply:
I've been recommended before never to manually set ErrorStatus - it's reserved for actual errors.  See also https://github.com/canonical/operator/blob/9664fd1ad43992e0aa3fb3ede9213f886eba61ac/ops/model.py#L1834-L1835 that suggests ops framework will block it if you manually set errorstatus.

Reply:
> I am concerned this will not work well with _on_collect_unit_status.

I wasn't sure how to handle this.  There are two main paths this condition could be hit in collect status:

1. the original install or server start failed - that should completely shadow anything set from collect status though, because errorstatus is higher priority, and besides it shouldn't get to the update-status hook unless the user cancels retrying the install or configure hooks.  In this case, error seems fine to raise, because we've got that unrecoverable condition.
2. something changed on the unit (a user or something else caused the snap to get uninstalled, or the service disabled, or something is crashing the service, etc.) - here is where update-status is the only place to get a timely warning about that.  Again error is probably most appropriate given the feedback we received from the charm listing review.

I can't find docs, but I guess if ops framework gracefully handles raising an error, it should work the same in collect_status?

Reply:
> Edit: and it's true, the charm failed on install hook immediately because it's missing cos-agent relation

Ah I just tested, and now I understand your comment.  We can't unconditionally assert that the service is active, because the charm doesn't have the service running unless cos-agent and keystone are ready.

Reply:
I think even if you remove "cos-agent and keystone are ready" check and start the exporter during config changed event, it still won't work. Because the collect status is run after install hook: install -> collect status (crash here) -> config changed -> collect status -> ...

Reply:
Updated - I went the other way - only check if the snap service is running if the charm expects it to be running (ie. keystone data is ready and cos-agent is related).

Reply:
ha, @jneo8 they just updated `ops` days ago so that adding `ErrorStatus` or `UnknownStatus` will crash the charm

Reply:
In the document 
https://ops.readthedocs.io/en/latest/#ops.CollectStatusEvent

> It is an error to call [add_status()](https://ops.readthedocs.io/en/latest/#ops.CollectStatusEvent.add_status) with an instance of [ErrorStatus](https://ops.readthedocs.io/en/latest/#ops.ErrorStatus) or [UnknownStatus](https://ops.readthedocs.io/en/latest/#ops.UnknownStatus).